### PR TITLE
Use our generated GTK bundle for travis

### PR DESCRIPTION
**Pull request by kugel-**
This enables travis to download and use the gtk bundles which we also use for the win32 installer images.

The last commit is just for testing. Do not merge. I expect we want to put the bundle on geany.org

### CODE REVIEW

**Comment by b4n** [comment]
99fb6d7847220c502dca08a5ffe1fb8ae314080a huh?  the whole point would be using a bundle that actually works so we know we really test something, otherwise my hack would work just fine too.  If the bundle doesn't work to corss-compile, it's probably useless.

Apart that, we ay want an option to still strip dev files in some situations.

@eht16

**Comment by eht16** [comment]
Nice idea.

However, like @b4n, I'm not so keen on keeping the development files in the created bundle. While is necessary for compilation, the main intention of the bundle creation was to provide a *runtime* environment. If at all, I think there should be a command line switch to toggle not removing development files. I don't think we want them into the installer.

About the location: we can host it easily on download.geany.org/contrib or download.geany.org/win32 or something like this.

**Comment by kugel-** [comment]
@b4n why add new requirements now? The travis result was never runtime tested so far why do you require it know?

Be aware that the bundle is compiled is with a different, much newer GCC. This includes libstdc++. I'm not sure it's officially supported to mix different compiler versions like this in the C++ space. The bundle is also not designed for cross compilation at all.

We use (and test) the bundle when there's a release coming up, I think that's sufficient.

@eht16 I don't care, however the development files are also useful when developing plugins without msys2 environment.

**Comment by b4n** [comment]
Well, the point of Travis is to see whether nothing broke.  If the build is totally fake, how can we trust the results?
And no, the previous bundle wasn't the one for the last release, but it was for older ones, and yes, I did test the builds.  I used the very same script on my machine to cross-compile several times, and tested the resulting binaries.
What I don't want is the Travis builds to break in two weeks not because we did something wrong but because it wasn't a reliable test but some fragile thing that happened to pass build.
That's why I didn't fancy my original idea of overwriting some part of the bundle with a newer one in a way I didn't even test produced a working executable.  I don't fancy either editing the bundle to strip out some flags just because it doesn't build with them, in a way that wasn't tested either.

**Comment by kugel-** [comment]
Can you please check if the binary works with the updated cross build script if you say you did so in the past? I got some linker error when trying to run it under wine.

**Comment by b4n** [comment]
I'll try tomorrow (not that I'd be very confident with it even if it did run with that flag stripping).
Though, don't you have a real Windows (virtual?) machine, as you seem to play with MSYS2 builds I would hope you do :)

**Comment by kugel-** [comment]
I only cross compiled in my ubuntu 12.04 container so far (to match travis); here I got the wine error. I didn't yet try the script on my actual Linux system with newer mingw-w64-gcc.

**Comment by codebrainz** [comment]
> however the development files are also useful when developing plugins without msys2 environment.

It seems reasonable to include all the needed header files as an option in the installer (ex. under the existing Development Files option) so people making plugins don't have to reverse engineer the G* stack we provide to find out what version of headers they need to match against which runtime libs, etc. I'm assuming it's just a bunch of headers and not something huge (relative to the .dlls).

**Comment by codebrainz** [comment]
> here I got the wine error

I wouldn't trust WINE to prove anything works or not, but what error did you get anyway?

File: scripts/cross-build-mingw.sh [hunk 1]
```diff
@@ -88,6 +88,8 @@ mkdir _deps
 fetch_and_unzip "$BUNDLE_ZIP" _deps
 # fixup the prefix= in the pkg-config files
 sed -i "s%^\(prefix=\).*$%\1$PWD/_deps%" _deps/lib/pkgconfig/*.pc
+# mingw-w64 doesn't know -pthread
+sed -i -e 's%-pthread%%g' _deps/lib/pkgconfig/*.pc
```
Comment (codebrainz) [inline scripts/cross-build-mingw.sh]: I think `-lpthread` works on Linux, MinGW and OSX GCC/glibc.

**Comment by eht16** [comment]
> It seems reasonable to include all the needed header files as an option in the installer (ex. under the existing Development Files option) so people making plugins don't have to reverse engineer the G* stack we provide to find out what version of headers they need to match against which runtime libs, etc. I'm assuming it's just a bunch of headers and not something huge (relative to the .dlls).
I don't think so. Since we don't do anything special with the GTK stack
we provide, there is no need to "reverse engineer" it. What we do is
installing packages with MSYS2 and remove some unnecessary files.
And necessary scripts for developing are in the repo as well as a howto
in the wiki.
To be complete in this regard, we would maybe even have to include the
toolchain into the installer which is overkill, IMO.

Btw, just the headers won't be enough. We would need to include the .a
files probably as well and they have some significant file size.

**Comment by codebrainz** [comment]
@eht16 by "reverse engineer" I meant plugin devs would have to figure out and download the version of G* headers which go with the specific version of G* dlls we ship in order to develop plugins. If we (optionally) shipped the G* headers along with the dlls we ship, all they would have to do is point their toolchain/compiler/buildsystem at our headers and dlls and they can make plugins, whether they use MSYS, Cygwin, MinGW, MSVC, ICC, etc. They should all interoperate as it's all plain C, including Scintilla which we use via a C API/ABI.

**Comment by codebrainz** [comment]
> Btw, just the headers won't be enough. We would need to include the .a files probably as well and they have some significant file size.

I think you can link to the .dlls directly but I could be wrong. If we ship c/c++ stdlib though, we'd need to include all of those c/c++ stdlib headers as well, and all of them in between there and G* (freetype, pango, pixman, etc.), which is not insignificant.

**Comment by kugel-** [comment]
You need to include headers, pkg-config files and some *.dll.a files for linking. The .dll.a files are small each.

**Comment by eht16** [comment]
du -chs */*.a for:
Geany's dependency bundle: 27 MB
Plugins' dependency bundle: 84 MB
I wouldn't call this that small.

Anyway, I still don't get the point shipping all those files is necessary for plugin developers. We have the howto on how to set up a full development environment on Windows. This is at least partially necessary to get a working toolchain. And at this point, building Geany from source is probably more easy than to fiddle with API/ABI compability of a binary Geany release, IMO.

Is this a real world use case or just a thought what users/plugin developers might do?
I personally would expect that people who are actually taking the pain to set up MSYS2 and everything around on Windows, are also willing to compile Geany from source which is way easier than building the plugins.

**Comment by kugel-** [comment]
The idea was to allow for plugin building without setting up msys2

**Comment by eht16** [comment]
With what toolchain then? I doubt ABI compability can be kept if a different toolchain would be used, especially MSVC.

**Comment by kugel-** [comment]
Am 17. Juni 2016 00:25:00 MESZ, schrieb "Enrico Tröger" <notifications@github.com>:
>With what toolchain then? I doubt ABI compability can be kept if a
>different toolchain would be used, especially MSVC.
>
>---
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/geany/geany/pull/1072#issuecomment-226631672

You can download prebuilt mingw binaries like always. Haven't checked binary compatibility though

**Comment by elextr** [comment]
> With what toolchain then? I doubt ABI compability can be kept if a different toolchain would be used, especially MSVC.

A user downloading binaries would have to know which toolchain they were built with and would be restricted to that toolchain and the same set of options as Enrico chose to build the Geany binary.  If they have a toolchain to build plugins then they can build Geany just as easily and be assured of compatibility.  And the source tarball has all the things needed to build plugins (except the toolchain).

**Comment by codebrainz** [comment]
> I doubt ABI compability can be kept if a different toolchain would be used, especially MSVC

I'm pretty sure it can with plain C, which the plugin API is (Scintilla's C++ code is hidden behind a C-only API). I believe most compilers (MinGW specifically) interoperate with Windows ABI/calling conventions. It's the same reason an application compiled with MinGW can use Win32 DLLs/runtime (ex. user32, shell32, etc) compiled with a different compiler (presumably MSVC), or why one can use GLib/GTK+ dlls compiled by other compilers with VisualStudio/MSVC. All that should be needed in addition to the .dlls we ship are the headers, there should be no need for .lib or .a stuff.

**Comment by kugel-** [comment]
@b4n Can we just use your combined bundle that works with cross toolchains and close this?

**Comment by b4n** [comment]
@kugel- yeah.  I tried to run both my combined versions, and

 * adding new glib on top of GTK2 bundle leads to missing newer versions of some stuff (*libintl* IIRC), so it's not very practical (we could add those additionally, sure)
 * combining the GTK3 bundle on top the GTK2 bundle lacks *libgcc_s_sjlj-1.dll* and *libstdc++-6*, but those are in the building toolchain, and adding those make it work (well, at least it runs).

I don't remember where you talked about it (IRC), but with the combined bundles, it has *libgcc_s_dw2-1.dll*, and we need to add *libgcc_s_sjlj-1.dll*.  I don't really know what it means to us, but it doesn't seem to be an actual runtime issue.  But as AFAIK we don't propagate any exceptions outside a single library (Scintilla), I guess it doesn't matter.

So, we could use the GTK3/GTK2 combination for Travis, it makes something that can be used -- well, at least run.

**Comment by b4n** [comment]
https://github.com/b4n/geany/tree/windows-cross-newer-glib

**Comment by b4n** [comment]
And as for including GTK dev files in the bundle: I have the same concerns as @eht16, I don't see a real value to it as users will require setting up a build environment anyway, and we document how to get that up with everything, not only GTK.

**Comment by codebrainz** [comment]
> I don't see a real value to it as users will require setting up a build environment anyway

The value is the exact same as the value added by the installer allowing to install Geany's development headers. Since there's not even that many compared to the GTK+ stack and C/C++ runtime, it makes even less sense to ship those (in the absence of all the other development headers).

An analogy is the value added by Debian repositories having `-dev` packages for the version of binary packages, rather than making users track down all of the correct headers for the version of each binary package.

In any case, for my own use I don't actually care, it just seems pointless to include some but not all development files needed to build plugins.

**Comment by elextr** [comment]
>  I don't see a real value to it as users will require setting up a build environment anyway,

+1

**Comment by kugel-** [comment]
You download prebuilt MingW gcc builds without setting up msys2. This plus dev files shipped by Geany would allow for more convenient plugin development that setting up msys2, IMO.

I don't feel strongly though.

**Comment by eht16** [comment]
A compromise could be to provide a seperate installer, containing everything necessary to start developing Geany and/or plugins on Windows.

But I personally don't want to include this in the normal, end-user installer. Nor do I want to create/maintain that seperate installer.

**Comment by codebrainz** [comment]
AFAIK, we do provide everything needed for someone to develop Geany plugins using their favourite toolchain (ie. the .dlls and Geany's headers), except they would have to figure out which GTK stack and C runtime versions we ship and track down the appropriate headers for those.

In any case, I'm not volunteering (for this next release) to make a script to pull in all the headers to the installer under the Development Files option, so I'm hardly in a position push for it :)

**Comment by kugel-** [comment]
The generated GTK bundle is compiled with an incompatible toolchian. It compiles and links fine but the resulting binary doesn't run (different, incompatible exception handling). This PR isn't required anymore as @b4n found a workaround to merge gtk3 and gtk2 bundle for newer glib on gtk2 builds (for travis).

**Pull request closed without merging**